File: src-tauri/src/backend/manager.rs
//! Backend Manager
//!
//! Manages the FastAPI backend as a subprocess.
//! Main coordinator that uses config, process, and health modules.
//! Mirrors electron/backend/manager.js

use std::path::PathBuf;
use std::process::Child;
use std::sync::Arc;
use std::io::{BufRead, BufReader};
use tokio::sync::Mutex;
use tokio::time::{Duration, interval};

use super::config::{
    detect_portable_mode, get_bind_host, get_data_dir, get_default_port,
    get_local_ip, get_network_settings, get_settings_path, NetworkSettings,
};
use super::health::{
    health_check, is_port_free, kill_zombie_processes, kill_processes_on_port,
    wait_for_ready,
};
use super::process::{
    force_kill_process, graceful_kill_process, kill_uvicorn_processes, spawn_backend,
};

/// Backend manager state
pub struct BackendManager {
    /// The backend process handle
    process: Arc<Mutex<Option<Child>>>,
    /// Portable data directory if in portable mode
    portable_data_dir: Option<PathBuf>,
    /// Port the backend is running on
    port: u16,
    /// Whether the app is packaged (production) or development
    is_packaged: bool,
    /// Restart attempt counter
    restart_attempts: Arc<Mutex<u32>>,
    /// Maximum restart attempts
    max_restart_attempts: u32,
    /// Whether the manager is shutting down
    shutting_down: Arc<Mutex<bool>>,
}

impl BackendManager {
    /// Create a new backend manager
    pub fn new(is_packaged: bool) -> Self {
        let portable_data_dir = detect_portable_mode();
        let is_portable = portable_data_dir.is_some();

        // Get port from settings or use default
        let network_settings = get_network_settings(portable_data_dir.as_ref());
        let default_port = get_default_port(is_portable);
        let port = network_settings.local_port.unwrap_or(default_port);

        log::info!(
            "[Backend] Mode: {} | Port: {}",
            if is_portable { "portable" } else { "system" },
            port
        );

        Self {
            process: Arc::new(Mutex::new(None)),
            portable_data_dir,
            port,
            is_packaged,
            restart_attempts: Arc::new(Mutex::new(0)),
            max_restart_attempts: 5,
            shutting_down: Arc::new(Mutex::new(false)),
        }
    }

    /// Check if running in portable mode
    pub fn is_portable(&self) -> bool {
        self.portable_data_dir.is_some()
    }

    /// Get the data directory
    pub fn get_data_dir(&self) -> PathBuf {
        get_data_dir(self.portable_data_dir.as_ref())
    }

    /// Get the settings path
    pub fn get_settings_path(&self) -> PathBuf {
        get_settings_path(self.portable_data_dir.as_ref())
    }

    /// Get network settings
    pub fn get_network_settings(&self) -> NetworkSettings {
        get_network_settings(self.portable_data_dir.as_ref())
    }

    /// Get the local IP address
    pub fn get_local_ip(&self) -> String {
        get_local_ip()
    }

    /// Get the port the backend is running on
    pub fn get_port(&self) -> u16 {
        self.port
    }

    /// Check if backend is running
    pub async fn is_running(&self) -> bool {
        self.process.lock().await.is_some()
    }

    /// Start the backend server
    pub async fn start(&self) -> Result<(), String> {
        // Check if already running
        if self.is_running().await {
            log::info!("[Backend] Already running");
            return Ok(());
        }

        // AGGRESSIVE cleanup - kill ALL uvicorn processes
        log::info!("[Backend] Cleaning up any existing uvicorn processes...");
        kill_uvicorn_processes();
        tokio::time::sleep(Duration::from_secs(1)).await;

        // Standard zombie kill on port
        if let Err(e) = kill_zombie_processes(self.port).await {
            log::warn!("[Backend] Warning: Could not free port: {}", e);
        }

        log::info!("[Backend] Starting server on port {}", self.port);

        let bind_host = get_bind_host(self.portable_data_dir.as_ref());
        let network_settings = self.get_network_settings();

        log::info!("[Backend] Binding to: {}", bind_host);
        if bind_host == "0.0.0.0" {
            let local_ip = self.get_local_ip();
            if network_settings.local_network_enabled {
                log::info!(
                    "[Backend] Local network access: enabled (http://{}:{})",
                    local_ip, self.port
                );
            }
            log::info!(
                "[Backend] Public access: {}",
                if network_settings.public_network_enabled { "enabled" } else { "disabled" }
            );
        }

        // Spawn the backend process
        let mut child = spawn_backend(
            &bind_host,
            self.port,
            self.is_packaged,
            self.portable_data_dir.as_ref(),
        ).map_err(|e| format!("Failed to spawn backend: {}", e))?;

        // Set up stdout/stderr logging in background
        let pid = child.id();

        if let Some(stdout) = child.stdout.take() {
            let reader = BufReader::new(stdout);
            std::thread::spawn(move || {
                for line in reader.lines().map_while(Result::ok) {
                    if !line.is_empty() {
                        log::info!("[Backend] {}", line);
                    }
                }
            });
        }

        if let Some(stderr) = child.stderr.take() {
            let reader = BufReader::new(stderr);
            std::thread::spawn(move || {
                for line in reader.lines().map_while(Result::ok) {
                    if !line.is_empty() {
                        if line.contains("ERROR") || line.contains("Exception") {
                            log::error!("[Backend Error] {}", line);
                        } else {
                            log::info!("[Backend] {}", line);
                        }
                    }
                }
            });
        }

        // Store the process
        *self.process.lock().await = Some(child);

        // Wait for backend to be ready
        wait_for_ready(self.port, 30000).await?;

        // Start health check task
        self.start_health_check();

        // Reset restart counter on successful start
        *self.restart_attempts.lock().await = 0;

        log::info!("[Backend] Started successfully (PID: {:?})", pid);
        Ok(())
    }

    /// Start periodic health checks
    fn start_health_check(&self) {
        let port = self.port;
        let process = Arc::clone(&self.process);
        let shutting_down = Arc::clone(&self.shutting_down);

        tokio::spawn(async move {
            let mut interval = interval(Duration::from_secs(30));

            loop {
                interval.tick().await;

                // Stop if shutting down
                if *shutting_down.lock().await {
                    break;
                }

                // Check if process exists
                if process.lock().await.is_some() {
                    let healthy = health_check(port).await;
                    if !healthy {
                        log::warn!("[Backend] Health check failed");
                    }
                } else {
                    // Process is gone, stop checking
                    break;
                }
            }
        });
    }

    /// Stop the backend server gracefully
    pub async fn stop(&self) -> Result<(), String> {
        *self.shutting_down.lock().await = true;

        let mut process_guard = self.process.lock().await;

        if process_guard.is_none() {
            log::info!("[Backend] Not running");
            return Ok(());
        }

        log::info!("[Backend] Initiating graceful shutdown...");

        let child = process_guard.as_mut().unwrap();
        let pid = child.id();

        // Send SIGINT first for clean uvicorn shutdown
        #[cfg(not(target_os = "windows"))]
        {
            log::info!("[Backend] Sending SIGINT...");
            graceful_kill_process(pid);
        }

        #[cfg(target_os = "windows")]
        {
            graceful_kill_process(pid);
        }

        // Wait for graceful shutdown with timeout
        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(10);

        loop {
            match child.try_wait() {
                Ok(Some(_)) => {
                    log::info!("[Backend] Stopped gracefully");
                    *process_guard = None;
                    return Ok(());
                }
                Ok(None) => {
                    if start.elapsed() > timeout {
                        break;
                    }
                    // Still running, escalate after 5 seconds
                    #[cfg(not(target_os = "windows"))]
                    if start.elapsed() > Duration::from_secs(5) {
                        log::info!("[Backend] Escalating to SIGTERM...");
                        let _ = child.kill();
                    }
                    tokio::time::sleep(Duration::from_millis(100)).await;
                }
                Err(e) => {
                    log::error!("[Backend] Error checking process status: {}", e);
                    break;
                }
            }
        }

        // Force kill if still running
        log::info!("[Backend] Graceful shutdown timeout, force killing...");
        self.force_kill_internal(&mut process_guard);

        Ok(())
    }

    /// Force kill the backend process
    fn force_kill_internal(&self, process_guard: &mut Option<Child>) {
        if let Some(child) = process_guard.as_mut() {
            let pid = child.id();
            force_kill_process(pid);
            let _ = child.kill();
        }
        *process_guard = None;

        // Also kill any zombie uvicorn processes
        kill_uvicorn_processes();
        kill_processes_on_port(self.port);
    }

    /// Force kill the backend (public)
    pub async fn force_kill(&self) {
        let mut process_guard = self.process.lock().await;
        self.force_kill_internal(&mut process_guard);
    }

    /// Restart the backend server
    pub async fn restart(&self) -> Result<(), String> {
        self.stop().await?;
        tokio::time::sleep(Duration::from_secs(1)).await;
        self.start().await
    }

    /// Perform health check
    pub async fn health_check(&self) -> bool {
        health_check(self.port).await
    }

    /// Check if port is free
    pub fn is_port_free(&self) -> bool {
        is_port_free(self.port)
    }
}

impl Drop for BackendManager {
    fn drop(&mut self) {
        // Synchronous cleanup on drop
        if let Ok(mut guard) = self.process.try_lock() {
            if guard.is_some() {
                self.force_kill_internal(&mut guard);
            }
        }
    }
}

File: electron/backend/manager.js
/**
 * Backend Manager
 * Manages the FastAPI backend as a subprocess
 * Main coordinator class that uses config, process, and health modules
 */
const {
  detectPortableMode,
  getDataDir,
  getSettingsPath,
  getNetworkSettings,
  getLocalIP,
  getBindHost
} = require('./config');

const {
  spawnBackend,
  forceKillProcess,
  gracefulKillProcess
} = require('./process');

const {
  healthCheck,
  waitForReady,
  isPortFree,
  getPortUser,
  killZombieProcesses,
  killProcessesOnPort
} = require('./health');

class BackendManager {
  constructor() {
    this.process = null;
    this.healthCheckInterval = null;
    this.restartAttempts = 0;
    this.maxRestartAttempts = 5;
    this.portableDataDir = null;

    // Detect portable mode on construction
    this.portableDataDir = detectPortableMode();

    // Different default ports: portable=8791, system=8790
    // This allows running both simultaneously without conflicts
    this.defaultPort = this.portableDataDir ? 8791 : 8790;

    // Read port from settings (user can override the default)
    const networkSettings = this.getNetworkSettings();
    this.port = networkSettings.local_port || this.defaultPort;
    console.log('[Backend] Mode:', this.portableDataDir ? 'portable' : 'system', '| Port:', this.port);
  }

  /**
   * Check if running in portable mode
   */
  isPortable() {
    return this.portableDataDir !== null;
  }

  /**
   * Get LocalBooru data directory (matches Python API location)
   */
  getDataDir() {
    return getDataDir(this.portableDataDir);
  }

  /**
   * Get settings.json path
   */
  getSettingsPath() {
    return getSettingsPath(this.portableDataDir);
  }

  /**
   * Load network settings from settings.json
   */
  getNetworkSettings() {
    return getNetworkSettings(this.portableDataDir);
  }

  /**
   * Get the local IP address
   */
  getLocalIP() {
    return getLocalIP();
  }

  /**
   * Determine the host to bind to based on network settings
   */
  getBindHost() {
    return getBindHost(this.portableDataDir);
  }

  /**
   * Get information about what process is using the port
   */
  getPortUser() {
    return getPortUser(this.port);
  }

  /**
   * Kill any zombie processes on our port
   * Throws PortConflictError if port cannot be freed
   */
  async killZombieProcesses() {
    return killZombieProcesses(this.port);
  }

  /**
   * Check if port is free
   */
  isPortFree() {
    return isPortFree(this.port);
  }

  /**
   * Start the backend server
   */
  async start() {
    if (this.process) {
      console.log('[Backend] Already running');
      return;
    }

    // Clean up zombie processes on OUR port only (not all uvicorn system-wide)
    // This allows multiple instances (portable + system) to coexist
    try {
      await this.killZombieProcesses();
    } catch (e) {
      // Log but don't fail - we'll try to start anyway and let uvicorn fail if port is busy
      console.error('[Backend] Warning: Could not free port:', e.message);
    }

    console.log('[Backend] Starting server on port', this.port);

    const bindHost = this.getBindHost();
    const networkSettings = this.getNetworkSettings();

    console.log('[Backend] Binding to:', bindHost);
    if (bindHost === '0.0.0.0') {
      const localIP = this.getLocalIP();
      console.log('[Backend] Local network access:', networkSettings.local_network_enabled ? `enabled (http://${localIP}:${this.port})` : 'disabled');
      console.log('[Backend] Public access:', networkSettings.public_network_enabled ? 'enabled' : 'disabled');
    }

    // Spawn the backend process
    this.process = spawnBackend({
      bindHost,
      port: this.port,
      portableDataDir: this.portableDataDir
    });

    // Handle stdout
    this.process.stdout.on('data', (data) => {
      const output = data.toString().trim();
      if (output) {
        console.log('[Backend]', output);
      }
    });

    // Handle stderr
    this.process.stderr.on('data', (data) => {
      const output = data.toString().trim();
      if (output) {
        // Uvicorn logs to stderr, so not all are errors
        if (output.includes('ERROR') || output.includes('Exception')) {
          console.error('[Backend Error]', output);
        } else {
          console.log('[Backend]', output);
        }
      }
    });

    // Handle exit
    this.process.on('exit', (code, signal) => {
      console.log(`[Backend] Exited with code ${code}, signal ${signal}`);
      this.process = null;

      // Auto-restart if crashed unexpectedly
      if (code !== 0 && this.restartAttempts < this.maxRestartAttempts) {
        this.restartAttempts++;
        console.log(`[Backend] Restarting (attempt ${this.restartAttempts}/${this.maxRestartAttempts})...`);
        setTimeout(() => this.start(), 2000);
      }
    });

    // Handle error
    this.process.on('error', (err) => {
      console.error('[Backend] Failed to start:', err);
      this.process = null;
    });

    // Wait for backend to be ready
    await waitForReady(this.port);

    // Start health check
    this.startHealthCheck();

    // Reset restart counter on successful start
    this.restartAttempts = 0;
  }

  /**
   * Wait for the backend to respond to health checks
   */
  async waitForReady(timeout = 30000) {
    return waitForReady(this.port, timeout);
  }

  /**
   * Check if backend is healthy and responding with valid content
   */
  healthCheck() {
    return healthCheck(this.port);
  }

  /**
   * Start periodic health checks
   */
  startHealthCheck() {
    this.healthCheckInterval = setInterval(async () => {
      if (this.process) {
        const healthy = await this.healthCheck();
        if (!healthy) {
          console.warn('[Backend] Health check failed');
        }
      }
    }, 30000); // Every 30 seconds
  }

  /**
   * Stop the backend server gracefully
   */
  async stop() {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }

    if (!this.process) {
      console.log('[Backend] Not running');
      return;
    }

    console.log('[Backend] Initiating graceful shutdown...');

    // First, try to notify the backend via a shutdown request
    // This gives FastAPI time to cleanup before SIGTERM
    try {
      const http = require('http');
      await new Promise((resolve, reject) => {
        const req = http.request({
          hostname: '127.0.0.1',
          port: this.port,
          path: '/health',  // Just check it's responding
          method: 'GET',
          timeout: 1000
        }, () => resolve());
        req.on('error', () => resolve());  // Ignore errors
        req.on('timeout', () => { req.destroy(); resolve(); });
        req.end();
      });
    } catch (e) {
      // Backend may already be unresponsive
    }

    return new Promise((resolve) => {
      // Give more time for graceful shutdown (thread pools, db connections, etc.)
      const timeout = setTimeout(() => {
        console.log('[Backend] Graceful shutdown timeout, force killing...');
        this.forceKill();
        resolve();
      }, 10000);  // Increased from 5s to 10s

      this.process.once('exit', () => {
        clearTimeout(timeout);
        this.process = null;
        console.log('[Backend] Stopped gracefully');
        resolve();
      });

      // Send SIGINT first (Ctrl+C equivalent) for cleaner uvicorn shutdown
      // Then SIGTERM if needed
      if (process.platform === 'win32') {
        // Windows: CTRL_C_EVENT doesn't work well with spawn, use taskkill
        gracefulKillProcess(this.process.pid);
      } else {
        // Unix: Send SIGINT first (cleaner uvicorn shutdown)
        console.log('[Backend] Sending SIGINT...');
        this.process.kill('SIGINT');

        // If still running after 5s, escalate to SIGTERM
        setTimeout(() => {
          if (this.process) {
            console.log('[Backend] Escalating to SIGTERM...');
            this.process.kill('SIGTERM');
          }
        }, 5000);
      }
    });
  }

  /**
   * Force kill the backend process AND any zombie uvicorn processes
   */
  forceKill() {
    // Kill our tracked process
    if (this.process) {
      forceKillProcess(this.process.pid);
      this.process = null;
    }

    // Kill anything on our port (not all uvicorn system-wide)
    killProcessesOnPort(this.port);
  }

  /**
   * Restart the backend server
   */
  async restart() {
    await this.stop();
    await this.start();
  }

  /**
   * Check if backend is running
   */
  isRunning() {
    return this.process !== null;
  }
}

module.exports = BackendManager;
